modules/projections/src/main/scala/ch/epfl/bluebrain/nexus/sourcing/projections/StreamSupervisor.scala

```diff
@@ -52,66 +52,45 @@ object StreamSupervisor {
   /**
     * Actor implementation that builds and manages a stream ([[RunnableGraph]]).
     *
-    * @param init   an initialization function to be run when the actor starts, or when the stream is restarted
-    * @param source an initialization function that produces a stream from an initial start value
+    * @param sourceF a source wrapped on the effect type F[_]
     */
-  class StreamSupervisorActor[F[_], A: ClassTag](init: F[A], source: A => Source[A, _])(implicit F: Effect[F])
+  class StreamSupervisorActor[F[_], A](sourceF: F[Source[A, _]])(implicit F: Effect[F])
       extends Actor
       with ActorLogging {
 
-    private val A                             = implicitly[ClassTag[A]]
     private implicit val as: ActorSystem      = context.system
     private implicit val ec: ExecutionContext = as.dispatcher
     //noinspection ActorMutableStateInspection
     private var state: Option[A] = None
 
-    private def initialize(): Unit = {
-      val logError: PartialFunction[Throwable, F[Unit]] = {
-        case err =>
-          log.error(err, "Failed on initialize function with error '{}'", err.getMessage)
-          F.raiseError(err)
-      }
-      val _ = init.map(Start).onError(logError).toIO.unsafeToFuture() pipeTo self
-    }
-
     override def preStart(): Unit = {
       super.preStart()
       initialize()
     }
 
-    private def buildStream(a: A): RunnableGraph[(UniqueKillSwitch, Future[Done])] = {
-      source(a)
-        .viaMat(KillSwitches.single)(Keep.right)
-        .map { latest =>
-          state = Some(latest)
-          latest
+    def initialize(): Unit =
+      buildStream()
+        .map { stream =>
+          val (killSwitch, doneFuture) = stream.run()
+          context.become(running(killSwitch))
```

Comment:
I think context.become is unsafe to be called from outside the message run loop.

Reply:
I changed to sending a message to self with the killSwitch